# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
%matplotlib inline
# print current directory
print(os.getcwd())

C:\Users\praka\Desktop\springboard\chapter6\guided_capstone


**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [16]:
df=pd.read_csv('data/step3.1_output.csv')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329 entries, 0 to 328
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               329 non-null    object 
 1   state              329 non-null    object 
 2   summit_elev        329 non-null    int64  
 3   vertical_drop      329 non-null    int64  
 4   trams              329 non-null    int64  
 5   fastEight          329 non-null    float64
 6   fastSixes          329 non-null    int64  
 7   fastQuads          329 non-null    int64  
 8   quad               329 non-null    int64  
 9   triple             329 non-null    int64  
 10  double             329 non-null    int64  
 11  surface            329 non-null    int64  
 12  Runs               329 non-null    float64
 13  TerrainParks       329 non-null    float64
 14  LongestRun_mi      329 non-null    float64
 15  SkiableTerrain_ac  329 non-null    float64
 16  Snow Making_ac     329 non

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [17]:
# creating dummies for state, drop first=True to get rid of dummy variable trap
dummies=pd.get_dummies(df['state'],drop_first=True) 
df=pd.concat([df.drop('state',axis=1),dummies],axis=1)
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 329 entries, 0 to 328
Data columns (total 58 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               329 non-null    object 
 1   summit_elev        329 non-null    int64  
 2   vertical_drop      329 non-null    int64  
 3   trams              329 non-null    int64  
 4   fastEight          329 non-null    float64
 5   fastSixes          329 non-null    int64  
 6   fastQuads          329 non-null    int64  
 7   quad               329 non-null    int64  
 8   triple             329 non-null    int64  
 9   double             329 non-null    int64  
 10  surface            329 non-null    int64  
 11  Runs               329 non-null    float64
 12  TerrainParks       329 non-null    float64
 13  LongestRun_mi      329 non-null    float64
 14  SkiableTerrain_ac  329 non-null    float64
 15  Snow Making_ac     329 non-null    float64
 16  daysOpenLastYear   329 non

In [18]:
df.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0          0          2     2       0       0  ...             0   
1          0          0     0       0       4  ...             0   
2          0          0     0       1       0  ...             0   
3          1          0     2       2       1  ...             0   
4          0          1     2       3       1  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 58 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [19]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.loc[:,'AdultWeekend']

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [20]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=10)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [21]:
#all first model set

from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [22]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [23]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

explained_variance_score(y_test, y_pred)

0.9229945224324895

In [24]:
mean_absolute_error(y_test, y_pred)

5.470868739089696

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [25]:
lm.intercept_

63.70996460498648

In [26]:
lm.coef_

array([ 5.09411776e+00,  1.22195241e+00,  2.58039458e-01,  1.33031090e-01,
        1.82190183e-02,  1.20432604e+00,  1.80954191e+00,  1.58341260e+00,
        9.44734884e-01, -7.07724302e-01, -7.22805600e-01,  2.96261756e-01,
       -1.07192885e+00, -1.27947514e+00,  1.10801196e-01,  2.83071560e-01,
       -3.24726506e-01, -2.18600368e-02,  1.91332387e+01, -5.56381395e-01,
        1.29642740e+00,  2.64895543e-02, -1.19638707e+00, -2.82454745e+00,
       -3.04044681e+00, -2.54377944e-01, -2.25652465e+00, -3.89937981e-01,
       -5.02961759e-01, -3.23514502e-01,  4.49331280e-01,  4.78212077e-01,
        1.79010715e+00, -7.61536375e-01, -7.44796738e-01, -3.66042022e-01,
       -2.68540775e+00, -1.91708107e+00,  6.29361263e-01,  3.77154856e-01,
       -2.64574594e+00, -1.84545840e-01,  1.13063008e+00, -1.19706329e+00,
       -1.56903059e+00,  6.16127932e-01,  2.29597738e-01, -9.34323864e-01,
        9.48050971e-01, -1.41320908e+00, -1.18909206e-01,  7.83256296e-01,
       -1.53385512e+00,  

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [27]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
#pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient'])

pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)

Coefficient
AdultWeekday    19.133239
summit_elev      5.094118
Colorado         3.040447
California       2.824547
Montana          2.685408
New Mexico       2.645746
Wyoming          2.387128
Idaho            2.256525
Nevada           1.917081
quad             1.809542

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [28]:
df_noState=pd.read_csv('data/step3_output.csv').drop('state', axis=1)
df_noState.head()

Name  summit_elev  vertical_drop  trams  fastEight  \
0       Alyeska Resort         3939           2500      1        0.0   
1  Eaglecrest Ski Area         2600           1540      0        0.0   
2     Hilltop Ski Area         2090            294      0        0.0   
3     Arizona Snowbowl        11500           2300      0        0.0   
4  Sunrise Park Resort        11100           1800      0        0.0   

   fastSixes  fastQuads  quad  triple  double  ...  SkiableTerrain_ac  \
0          0          2     2       0       0  ...             1610.0   
1          0          0     0       0       4  ...              640.0   
2          0          0     0       1       0  ...               30.0   
3          1          0     2       2       1  ...              777.0   
4          0          1     2       3       1  ...              800.0   

   Snow Making_ac  daysOpenLastYear  yearsOpen  averageSnowfall  AdultWeekday  \
0           113.0             150.0       60.0            669.0          65.0   
1            60.0              45.0       44.0            350.0          47.0   
2            30.0             150.0       36.0             69.0          30.0   
3           104.0             122.0       81.0            260.0          89.0   
4            80.0             115.0       49.0            250.0          74.0   

   AdultWeekend  projectedDaysOpen  NightSkiing_ac  clusters  
0          85.0              150.0           550.0         1  
1          53.0               90.0             0.0         1  
2          34.0              152.0            30.0         1  
3          89.0              122.0             0.0         0  
4          78.0              104.0            80.0         0  

[5 rows x 24 columns]

In [29]:
from sklearn import preprocessing

X = df_noState.drop(['Name','AdultWeekend'], axis=1)
y =df_noState.loc[:,'AdultWeekend']
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)

  
from sklearn.model_selection import train_test_split
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=10)

In [30]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

In [31]:
explained_variance_score(y_test, y_pred)

0.9305157642499965

In [32]:
mean_absolute_error(y_test, y_pred)

5.278311325467745

In [33]:
lm.intercept_

63.9239903812998

In [34]:
lm.coef_

array([-1.42589103,  2.67848445, -0.55125978, -0.19011728,  0.67584376,
        1.25644254,  1.84818008,  1.61396802,  0.70229384, -0.84488911,
        0.39082984, -0.81271546, -0.50355464, -2.08427197,  0.08941063,
        1.5477524 , -0.30640925, -0.62364471, 19.2540485 , -1.21686034,
        1.18269699, -0.61554582])

In [35]:
coeff_df=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False)
coeff_df

Coefficient
AdultWeekday         19.254048
vertical_drop         2.678484
SkiableTerrain_ac     2.084272
quad                  1.848180
triple                1.613968
daysOpenLastYear      1.547752
summit_elev           1.425891
fastQuads             1.256443
projectedDaysOpen     1.216860
NightSkiing_ac        1.182697
surface               0.844889
TerrainParks          0.812715
double                0.702294
fastSixes             0.675844
averageSnowfall       0.623645
clusters              0.615546
trams                 0.551260
LongestRun_mi         0.503555
Runs                  0.390830
yearsOpen             0.306409
fastEight             0.190117
Snow Making_ac        0.089411

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [36]:
df_no=pd.read_csv('data/step3_output.csv').drop(['state','summit_elev', 'vertical_drop', 'averageSnowfall','yearsOpen','daysOpenLastYear','projectedDaysOpen'], axis=1)
df_no.head()


Name  trams  fastEight  fastSixes  fastQuads  quad  triple  \
0       Alyeska Resort      1        0.0          0          2     2       0   
1  Eaglecrest Ski Area      0        0.0          0          0     0       0   
2     Hilltop Ski Area      0        0.0          0          0     0       1   
3     Arizona Snowbowl      0        0.0          1          0     2       2   
4  Sunrise Park Resort      0        0.0          0          1     2       3   

   double  surface  Runs  TerrainParks  LongestRun_mi  SkiableTerrain_ac  \
0       0        2  76.0           2.0            1.0             1610.0   
1       4        0  36.0           1.0            2.0              640.0   
2       0        2  13.0           1.0            1.0               30.0   
3       1        2  55.0           4.0            2.0              777.0   
4       1        0  65.0           2.0            1.2              800.0   

   Snow Making_ac  AdultWeekday  AdultWeekend  NightSkiing_ac  clusters  
0           113.0          65.0          85.0           550.0         1  
1            60.0          47.0          53.0             0.0         1  
2            30.0          30.0          34.0            30.0         1  
3           104.0          89.0          89.0             0.0         0  
4            80.0          74.0          78.0            80.0         0

In [37]:
from sklearn import preprocessing

X = df_no.drop(['Name','AdultWeekend'], axis=1)
y =df_no.loc[:,'AdultWeekend']
scaler = preprocessing.StandardScaler().fit(X)
X_scaled=scaler.transform(X)
  
from sklearn.model_selection import train_test_split
y = y.ravel()
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=10)

In [38]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error

lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

y_pred = model.predict(X_test)

In [39]:
explained_variance_score(y_test, y_pred)

0.9285908424385263

In [40]:
mean_absolute_error(y_test, y_pred)

5.380319651483905

In [41]:
lm.intercept_

63.893940070535244

In [42]:
lm.coef_

array([ 6.48737706e-02, -1.13227076e-03,  6.25164084e-01,  1.83095806e+00,
        1.93829957e+00,  1.43501713e+00,  7.23349858e-01, -7.87839872e-01,
        5.22012240e-01, -8.62003801e-01,  1.08186161e-03, -2.46749586e+00,
        1.44651439e-01,  1.94280464e+01,  1.29023606e+00, -6.82073129e-01])

In [43]:
coeff_df=pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending=False).head(10)
coeff_df

Coefficient
AdultWeekday         19.428046
SkiableTerrain_ac     2.467496
quad                  1.938300
fastQuads             1.830958
triple                1.435017
NightSkiing_ac        1.290236
TerrainParks          0.862004
surface               0.787840
double                0.723350
clusters              0.682073

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | 0.928 |5.22  |-|
| Model 2. | 0.93|5.27 |'state'|
| Model 3. | 0.8167 |6.017 |'state','summit_elev', 'vertical_drop', 'averageSnowfall', 'yearsOpen', 'daysOpenLastYear', 'projectedDaysOpen'|

Model Selection:

**Model 3**- This excludes inherent features that the resort cannot manage.